### PR TITLE
Add support for ICL QA tasks and generation during evaluation with `HuggingFaceModel`

### PR DESCRIPTION
**Pull request by dakinggg**
# What does this PR do?
This PR does a few things. The main purpose of the PR is to add support for In Context Learning evaluation of generative tasks (i.e. open ended QA). In service of this,
- adds a `generate` capability to `HuggingFaceModel`, and calls into `generate` during `eval_forward` on batches with `mode=generate`
- adds a `InContextLearningQADataset`, which supports a dataset format with a question, answer, and answer aliases
- adds a `InContextLearningQAAccuracy` metric which takes text as input, normalizes it, and checks if the predicted text is a prefix for any of the answer aliases
- fixes a bug with all ICL datasets where if the dataset file is attempted to be retrieved on every rank instead of just local rank 0

TODOs:
- [x] Add unit tests for the new `HuggingFaceModel` capabilities
- [x] update all the relevant documentation
- [x] One final GPT neo run computing multiple ICL metrics, just to be sure. https://wandb.ai/mosaic-ml/daniel-debug/runs/ha2p5ih2?workspace=user-danielking

Points of comparison that the metric computation is reasonable (the eleuther eval harness doesn't appear to do the text normalization):
- eleuther eval harness with neo-2.7b limit 1000, 5-shot: 0.115
- ours neo 2.7b full, 5-shot: 0.17
- llama 7b after ~300b tokens, 64 shot?: 0.45 *evaluation procedure not open source
- eleuther eval harness with opt-6.7b limit 1000, 5-shot: 0.259
- ours opt-6.7b full, 5-shot: 0.3

Limitations:
- all of the ICL tasks are not really set up for encoder-decoder. This one is not different. We can add support to all of them at the same time in the future.

# What issue(s) does this change relate to?
[CO-1831](https://mosaicml.atlassian.net/browse/CO-1831)

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

[CO-1831]: https://mosaicml.atlassian.net/browse/CO-1831?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by bmosaicml — APPROVED** [review]
Great work Daniel! I know this alias thing is sued for TriviaQA. Are there any other ICL datasets you expect this to be useful for?

File: tests/models/test_hf_model.py
Comment (samhavens) [inline tests/models/test_hf_model.py:668]: ooof

File: composer/models/huggingface.py
Comment (samhavens) [inline composer/models/huggingface.py:428]: why these but not e.g. top_p?

Comment (samhavens) [reply composer/models/huggingface.py:428]: I assume we can pass that with kwargs? I am just curious why these are explicitly not in kwargs

**Review by samhavens — APPROVED** [review]
🚀

File: composer/datasets/in_context_learning_evaluation.py
Comment (mvpatel2000) [inline composer/datasets/in_context_learning_evaluation.py:61]:
```suggestion
        raise ValueError(f"Unknown padding_side {padding_side}. padding_side must be either 'left' or 'right'")
```

File: composer/datasets/in_context_learning_evaluation.py
Comment (mvpatel2000) [inline composer/datasets/in_context_learning_evaluation.py:120]: Currently, we use `dist.local_rank_zero_download_and_wait` to avoid dist barrier. However, if the download fails, this results in a deadlock -> NCCL timeout which is pretty nasty. What if we instead added a context that runs only local rank 0 and then uses a signal file (ensuring the path to signal file exists or use `/tmp/some_signal_file`)? This way we avoid nasty deadlocks with download failures

File: composer/datasets/in_context_learning_evaluation.py
Comment (mvpatel2000) [inline composer/datasets/in_context_learning_evaluation.py:173]: Are you serious? that's absurd...

File: composer/datasets/in_context_learning_evaluation.py
Comment (mvpatel2000) [inline composer/datasets/in_context_learning_evaluation.py:177]:
```suggestion
            # If the preamble is empty then this will be a 0-length list, unless the tokenizer adds special tokens to empty strings (e.g. OPT tokenizer)
```

File: composer/datasets/in_context_learning_evaluation.py
Comment (mvpatel2000) [inline composer/datasets/in_context_learning_evaluation.py:179]:
```suggestion
            # If there is an EOS token added, we need to remove it so it is not in the middle of the prompt
```

File: composer/metrics/nlp.py
Comment (mvpatel2000) [inline composer/metrics/nlp.py:358]: Can we add + epsilon (and make epsilon an argument) just to be safe? iirc other metrics implementations do this? 

I'm also fine erroring 🤷

File: composer/models/huggingface.py
Comment (mvpatel2000) [inline composer/models/huggingface.py:448]: "We have no plans to support this at the moment"  

F

I would recommend emitting a waring saying we are doing a dummy implementation, this will add some latency.

File: composer/models/huggingface.py
Comment (dakinggg) [reply composer/models/huggingface.py:428]: Yeah this is a fair question. I called out the really important ones explicitly for clarity...but there isn't any particular reason to do this. Open to putting everything in kwargs. Max length is extremely critical in order for this to work with fsdp, but the others are not.

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [reply composer/datasets/in_context_learning_evaluation.py:120]: I'm pretty sure if the file doesn't exist it crashes, but will double check.

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [reply composer/datasets/in_context_learning_evaluation.py:173]: 🤷‍♂️ not sure if it's a thing that is consistent across models, or if I was doing something else wrong. I should probably try to double check this before merging.

File: composer/models/huggingface.py
Comment (dakinggg) [reply composer/models/huggingface.py:448]: The latency is _extremely_ minor. This is one call to forward (and only once), on one token of input. Whereas generate is num_tokens calls to forward. Maybe worth adding a warning anyway. Also this reminded me that I need to check if this works for encoder decoders too.

File: composer/models/huggingface.py
Comment (samhavens) [reply composer/models/huggingface.py:428]: `num_beams` feels less important to me than `top_p`, since nucleus sampling is considered better lately. And if we expose `num_beams` I feel like we should include `no_repeat_ngram_size`.

I think I am in favor of explicitly exposing all the generation parameters. One benefit to this is that we can **check if the user sets `penalty_alpha` to a non-falsey value; if so I am pretty sure we need to error, since I don't think we support contrastive decoding** currently (we will need to conform our forward signature to huggingface standards; we need to support [output_hidden_states](https://github.com/huggingface/transformers/blob/v4.27.1/src/transformers/generation/utils.py#L1834) and [passing around the KV cache](https://github.com/huggingface/transformers/blob/v4.27.1/src/transformers/generation/utils.py#L1858-L1862)

Comment (dakinggg) [reply composer/models/huggingface.py:428]: There are kind of a lot of generate parameters... https://huggingface.co/docs/transformers/v4.27.1/en/main_classes/text_generation#transformers.GenerationConfig. I don't think _we_ (being `HuggingFaceModel`) need to error on any generation parameters, because the call to generate happens on the underlying HF model, and presumably HF errors if you do stuff that is not allowed. We (being `MosaicGPT`) may need some error handling for things that aren't supported yet in the `MosaicGPT` model. Also, if there are any args that we need to check here for some reason, we can just pop them off of. Ok, what if I just don't specially expose any args, and defer to the HF docs?

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [reply composer/datasets/in_context_learning_evaluation.py:173]: Yeah, that is definitely the reason, I'm just not completely sure if the prompt _not_ ending in a space is a thing that all models prefer.

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [reply composer/datasets/in_context_learning_evaluation.py:120]: confirmed that it just fails cleanly if the file does not exist.

Comment (dakinggg) [reply composer/datasets/in_context_learning_evaluation.py:120]: adding the context you suggested is a good general idea, but going to skip for this pr

File: composer/metrics/nlp.py
Comment (dakinggg) [reply composer/metrics/nlp.py:358]: This is what you get if you call compute before update (all of our metrics have this property)
```
In [2]: from composer.metrics import MaskedAccuracy

In [3]: acc = MaskedAccuracy()

In [4]: acc.compute()
/workdisk/danielking/miniconda3/envs/examples-llm/lib/python3.10/site-packages/torchmetrics/utilities/prints.py:36: UserWarning: The ``compute`` method of metric MaskedAccuracy was called before the ``update`` method which may lead to errors, as metric states have not yet been updated.
  warnings.warn(*args, **kwargs)
Out[4]: tensor(nan)
```

Comment (dakinggg) [reply composer/metrics/nlp.py:358]: So I will leave it

File: composer/models/huggingface.py
Comment (dakinggg) [reply composer/models/huggingface.py:448]: fixed for t5. skipping the warning because this is so minor compared to the latency of generate.

File: composer/models/huggingface.py
Comment (dakinggg) [reply composer/models/huggingface.py]: ok i dropped all the explicit args and left a note in the docstring about where to find them in the HF docs.

File: composer/datasets/in_context_learning_evaluation.py
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:24]: padding side right does not work for decoder archs

**Comment by dakinggg** [comment]
@mvpatel2000 requested rereview for the changes to setup.py and data_spec.py

**Review by bandish-shah — APPROVED** [review]
LGTM

**Pull request merged**
